salt/metalk8s/volumes/prepared/files/sparse_volume_cleanup.py

```diff
@@ -0,0 +1,158 @@
+    sparse_file_path = os.path.join(SPARSE_FILES_DIR, volume_id)
+    if not os.path.isfile(sparse_file_path):
+        die(
+            "Sparse file {} does not exist for volume '{}'".format(
+                sparse_file_path, volume_id
+            ),
+            exit_code=errno.ENOENT
+        )
+
+
+def cleanup(volume_id):
+    """Clean up a sparse volume's loop device.
+
+    This function will bail out if a corresponding loop device cannot be
+    found.
+    """
+    device_path = find_device(volume_id)
+
+    print("Detaching loop device '{}' for volume '{}'".format(
+        device_path, volume_id
+    ))
+
+    device_handle = os.open(device_path, os.O_RDONLY)
+    try:
+        fcntl.ioctl(device_handle, LOOP_CLR_FD, 0)
+    except IOError as exn:
+        if exn.errno != errno.ENXIO:
+            die("Unexpected error when trying to free device {}: {}".format(
+                device_path, str(exn)
+            ))
+        print("Device already freed")
+    finally:
+        os.close(device_handle)
+
+    print("Loop device for volume '{}' was successfully detached".format(
+        volume_id
+    ))
+
+
+def main():
+    """Main "routine" of this script."""
+    args = parse_args()
+
+    check_sparse_file(args.volume_id)
+    cleanup(args.volume_id)
+
+
+# Helpers {{{
+def die(message, exit_code=1):
+    """Print a message to the standard error stream, and exit."""
+    print(message, file=sys.stderr)
+    sys.stderr.flush()
+    sys.exit(exit_code)
+
+
+def device_by_uuid(volume_id):
+    """Find a device by file-system UUID."""
+    path_by_uuid = os.path.join("/dev/disk/by-uuid", volume_id)
+    if not os.path.exists(path_by_uuid):
+        return None
+
+    return os.path.realpath(path_by_uuid)
+
+
+def device_from_part_uuid(volume_id):
+    """Find a device from its first partition UUID."""
+    path_by_partuuid = os.path.join("/dev/disk/by-partuuid", volume_id)
+    if not os.path.exists(path_by_partuuid):
+        return None
+
+    partition_name = os.path.basename(os.path.realpath(path_by_partuuid))
+    device_link = os.path.join("/sys/class/block", partition_name, os.pardir)
+    device_name = os.path.basename(os.path.realpath(device_link))
+    return os.path.join("/dev", device_name)
+
+
+def find_device(volume_id):
+    """Find the device provisioned for this volume ID.
+
+    SparseLoopDevice volumes are either:
+    - formatted, with the volume ID set in the file-system annotations
+    - raw, in which case the device is partitioned and the first partition
+      is labeled with the volume ID
+    """
+    for _get_device in [device_by_uuid, device_from_part_uuid]:
+        device_path = _get_device(volume_id)
+        if device_path is not None:
+            break
+    else:
+        die(
+            "Device for volume '{}' was not found".format(volume_id),
+            exit_code=errno.ENOENT,
+        )
+
+    if not is_loop_device(device_path):
+        die(
```

Comment:
minor: instead of having a `die` function, maybe raising an exception (+ global try/catch around `main()`) would be clearer (more idiomatic at least).
This (global try/catch) would also handle the errors you forgot/ignored on purpose (failure of `os.stat` and `os.open` for instance).

Not sure it's worth it though, it's your call.

Reply:
It's a good suggestion, and not too much work, I'll handle it.

Reply:
Did something really basic, should do the trick :slightly_smiling_face: